apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/samhsa/backfill/AbstractSamhsaBackfill.java

```diff
@@ -0,0 +1,268 @@
+package gov.cms.bfd.pipeline.sharedutils.samhsa.backfill;
+
+import gov.cms.bfd.pipeline.sharedutils.SamhsaUtil;
+import gov.cms.bfd.pipeline.sharedutils.TransactionManager;
+import gov.cms.bfd.pipeline.sharedutils.model.TableEntry;
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.NoResultException;
+import jakarta.persistence.Query;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import org.apache.commons.text.StringSubstitutor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract class to backfill the SAMHSA tags. This will iterate through each of the tables, pull
+ * any claims that don't already have SAMHSA tags, and checks them for SAMHSA codes, using
+ * SamhsaUtil.
+ */
+public abstract class AbstractSamhsaBackfill {
+  /** The Logger. */
+  private static final Logger LOGGER = LoggerFactory.getLogger(AbstractSamhsaBackfill.class);
+
+  /**
+   * Query to retrieve a list of claims objects, ignoring claims that already have SAMHSA tags. Will
+   * start at a given claim id, and limit the results to a given limit.
+   */
+  protected String QUERY_WITH_STARTING_CLAIM =
```

Comment:
I think my main concern with this is that we could potentially miss claims if new data is being updated while this is running since PAC claim IDs are not monotonic and older claims could be updated after we've already processed them. If we can be sure this runs in under a day or two, perhaps we can just pause the other pipelines while this runs.

Reply:
The other option would be to key off of `last_updated`, but that would require adding another index to all of the claim tables.

Reply:
I thought this would be a problem at first, as well, but then I realized that any new data that comes in will be processed by the RDA/CCW pipeline (which makes a call to SamhsaUtil before the claims are written).

Reply:
Ah yeah, very true. Should be fine then.